### DISCUSSION TITLE
CAT wallet showing CAT puzzle hash in incoming transactions

### DISCUSSION THREAD

**Discussion by ojura** [Chia-Network/chia-blockchain#9343] [Support]
The light wallet shows the CAT puzzle hash for incoming transactions, and not the real puzzle hash of the underlying standard chia wallet.

For example, one of my Chia receive addresses is `xch1....3hlnh`, and when you wrap that puzzle in a CAT puzzle (with the Spacebucks tail `0x78ad32a8c9ea70f27d73e9306fc467bab2a6b15b30289791e37ab6e8612212b1`), you get a puzzle hash `xch1...6sj34t68`.

This is how it looks like when I got some Spacebucks from their faucet:
![image](https://user-images.githubusercontent.com/9062353/142840491-c1093dce-f1ec-4536-93c0-1839004ffaf4.png)

However, an uninformed user may try to use this as an address for sending the CAT. The UI would suggest that everything is fine:

![image](https://user-images.githubusercontent.com/9062353/142840800-e535cc4c-cab9-4437-b723-f6947aabd425.png)

But this actually sent the funds to `xch1vnk6f5tm30y7zmyd0v2g2s9xt8fuvm894u82f4l560ttllv7ltsq6ry9u7`, which is a puzzle where we have a second outer Spacebucks CAT layer wrapping the original CAT in turn wrapping the original standard wallet puzzle. Of course, the wallet has no understanding of this, so it doesn't display this in the balance. 

In fact, can this coin even be spent? Will the inner CAT puzzle be messed up by the outer CAT layer?

**Comment by ojura** [comment]
Referencing question #9339, where another user unknowingly stumbled upon this.

I think this is worthy of a bug report, so opening https://github.com/Chia-Network/chia-blockchain/issues/9345 as well.

**Comment by Yostra** [comment]
Just to explain what's going on for everyone:
When you ask someone to send you a CAT you give them your inner puzzle, default is standard puzzle. 
When we receive CAT coin we display CAT wrapped version of standard puzzle as `to address`. 
I guess we could make this user friendlier and display  inner puzzle in UX.

**Reply by trgarrett** [reply]
Thanks for the explanation. It’s clear to me now what happened, but I do think the UX could be improved. Burning, or at least making it difficult to access, a few memecoins isn’t a game-changer for anyone, but if the more technical early adopters fall into this, this will definitely be a widespread challenge for mass adoption.

Since all coins and spends are based on puzzles, I think I might go one further than the UX change and question why a send to an address behaves differently enough with CATs to show up differently on the block explorers. The only advantage I can see is to keep the block explorers and wallets that don’t understand CAT from blindly tallying the address holdings as XCH. Maybe that prevents enough problems to make it worth it.

Without good offline wallet support in Chia, I know many users will try to leverage the block explorer exactly as I did and fall into the same predicament.

**Reply by BekzhanOmirzak** [reply]
Could you help how it is done?

**Comment by Yostra** [comment]
We should also have a "more details" page in ui for TXs so that we can display full puzzle also. This is useful if you want to check TXs on a explorer.

**Comment by ojura** [comment]
In that case, also display the outer puzzle for outgoing TX in _more details_?
Currently the inner puzzle is shown for outgoing, while the outer is shown for incoming transactions. This asymmetry is a bit confusing.
Also maybe include a warning not to use the outer puzzles for sending.

Do you have an idea if it is possible to spend coins which have been double CAT wrapped, or are they burned? I see this question coming up a lot as a result of these mistakes.

**Comment by ojura** [comment]
I don't think that these double wrapped CATs are able to be unwrapped to be regular CATs. 
https://github.com/Chia-Network/chia-blockchain/blob/d6441e8e05f80f112da8bd49e113ae28ec765fa6/chia/wallet/puzzles/cat.clvm#L163

Thus spending a coin which has two layers of CAT is going to wrap the destination puzzle hash twice, and you'll get the same monster again.

Perhaps the CAT puzzle could be tweaked to examine if the inner layer is also a CAT of the same type, and then to skip the second wrap. However this would require changing the CAT standard and breaking compatibility with current CATs.

**Comment by ojura** [comment]
Just an update: spending doubly wrapped coins **is** possible, albeit in a very limited fashion. The result of the spend can only be another doubly wrapped coin, which significantly restricts their use (e.g. combining them with regular CATs to send some amount of CAT, and receive back the change). They are also not recognized by the wallet, and spending them currently requires manually crafting a special spend bundle.

PR #9422 seeks to address this by introducing a mechanism into the CAT puzzle which will allow us to unwrap the multiply wrapped coin back into a regular CAT.
